rules/opentitan/cc.bzl

```diff
@@ -0,0 +1,329 @@
+        disassembly = disassembly,
+        mapfile = mapfile,
+    )
+    return provides, signed
+
+def _opentitan_binary(ctx):
+    providers = []
+    default_info = []
+    groups = {}
+    for exec_env in ctx.attr.exec_env:
+        exec_env = exec_env[ExecEnvInfo]
+        name = _binary_name(ctx, exec_env)
+        deps = [exec_env.lib] + ctx.attr.deps
+        provides, signed = _build_binary(ctx, exec_env, name, deps)
+        providers.append(exec_env.create_provider(
+            ctx,
+            exec_env,
+            **provides
+        ))
+        default_info.append(provides["default"])
+        groups.update(_as_group_info(exec_env.exec_env, signed))
+        groups.update(_as_group_info(exec_env.exec_env, provides))
+
+    providers.append(DefaultInfo(files = depset(default_info)))
+    providers.append(OutputGroupInfo(**groups))
+    return providers
+
+common_binary_attrs = {
+    "srcs": attr.label_list(
+        allow_files = True,
+        doc = "The list of C and C++ files that are processed to create the target.",
+    ),
+    "deps": attr.label_list(
+        providers = [CcInfo],
+        doc = "The list of other libraries to be linked in to the binary target.",
+    ),
+    "linker_script": attr.label(
+        providers = [CcInfo],
+        doc = "Linker script for linking this binary",
+    ),
+    "copts": attr.string_list(
+        doc = "Add these options to the C++ compilation command.",
+    ),
+    "defines": attr.string_list(
+        doc = "List of defines to add to the compile line.",
+    ),
+    "local_defines": attr.string_list(
+        doc = "List of defines to add to the compile line.",
+    ),
+    "includes": attr.string_list(
+        doc = "List of include dirs to be added to the compile line.",
+    ),
+    "linkopts": attr.string_list(
+        doc = "Add these flags to the C++ linker command.",
+    ),
+    "naming_convention": attr.string(
+        doc = "Naming convention for binary artifacts.",
+        default = "{name}_{exec_env}",
+    ),
+    "_cleanup_script": attr.label(
+        allow_single_file = True,
+        default = "@//rules/scripts:expand_tabs.sh",
+        doc = "Cleanup script for the disassembly",
+    ),
+}
+
+opentitan_binary = rv_rule(
+    implementation = _opentitan_binary,
+    attrs = dict(common_binary_attrs.items() + {
+        "exec_env": attr.label_list(
+            providers = [ExecEnvInfo],
+            doc = "List of exeuction environments for this target.",
+        ),
+        "_cc_toolchain": attr.label(default = Label("@bazel_tools//tools/cpp:current_cc_toolchain")),
+    }.items()),
+    fragments = ["cpp"],
+    toolchains = ["@rules_cc//cc:toolchain_type"],
+)
+
+def _opentitan_test(ctx):
+    exec_env = ctx.attr.exec_env[ExecEnvInfo]
+    if len(ctx.attr.srcs) == 1 and len(ctx.attr.deps) == 0 and exec_env.get_provider(ctx.attr.srcs[0]):
+        p = exec_env.get_provider(ctx.attr.srcs[0])
+    else:
+        name = _binary_name(ctx, exec_env)
```

Comment:
Why does this branch reimplements the inner loop of `_opentitan_binary` versus just called `_opentitan_binary`?

Reply:
`_opentitan_binary` expects to deal with a list of `exec_env` whereas the test rule expects to have only one `exec_env`.  The binary rule also produces a bunch of output providers that are not relevant to the test rule.